core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -577,13 +582,17 @@ private[deploy] class Worker(
     val shouldCleanup = finishedApps.contains(id) && !executors.values.exists(_.appId == id)
     if (shouldCleanup) {
       finishedApps -= id
-      appDirectories.remove(id).foreach { dirList =>
-        logInfo(s"Cleaning up local directories for application $id")
-        dirList.foreach { dir =>
-          Utils.deleteRecursively(new File(dir))
+      cleanupApplicationThreadExecutor.submit(new Runnable {
```

Comment:
That won't become an issue. The worst case is there will be some pending tasks in the queue of `cleanupThreadExecutor`. Considering the number of applications is not huge, it won't be an issue.